deeplake/integrations/mmdet/mmdet_.py

```diff
@@ -0,0 +1,309 @@
+from mmcv.utils import build_from_cfg, Registry
+from mmdet.datasets.builder import PIPELINES
+from mmdet.datasets.pipelines import Compose
+from mmcv.parallel import collate
+from functools import partial
+
+from mmdet.core import BitmapMasks
+import albumentations as A
+import deeplake as dp
+from deeplake.util.warnings import always_warn
+
+
+def _coco_2_pascal(boxes):
+    # Convert bounding boxes to Pascal VOC format and clip bounding boxes to make sure they have non-negative width and height
+    return np.stack(
+        (
+            boxes[:, 0],
+            boxes[:, 1],
+            boxes[:, 0] + np.clip(boxes[:, 2], 1, None),
+            boxes[:, 1] + np.clip(boxes[:, 3], 1, None),
+        ),
+        axis=1,
+    )
+
+
+rand_crop = A.Compose(
+    [
+        A.RandomSizedBBoxSafeCrop(width=128, height=128, erosion_rate=0.2),
+    ],
+    bbox_params=A.BboxParams(
+        format="pascal_voc",
+        label_fields=["labels", "bbox_ids"],
+        min_area=25,
+        min_visibility=0.6,
+    ),
+)
+
+
+def _find_tensor_with_htype(ds: dp.Dataset, htype: str):
+    tensors = [k for k, v in ds.tensors.items() if v.meta.htype == htype]
+    if not tensors:
+        raise ValueError("No tensor found with htype='{htype}'")
+    t = tensors[0]
+    if len(tensors) > 1:
+        always_warn(f"Multiple tensors with htype='{htype}' found. choosing '{t}'.")
+    return t
+
+
+def transform(
+    sample_in,
+    images_tensor: str,
+    masks_tensor: str,
+    boxes_tensor: str,
+    labels_tensor: str,
+    pipeline: Callable,
+):
+    img = sample_in[images_tensor]
+    masks = sample_in[masks_tensor]
+    bboxes = sample_in[boxes_tensor]
+    labels = sample_in[labels_tensor]
+
+    img = img[..., ::-1]
+    if img.shape[2] == 1:
+        img = np.repeat(img, 3, axis=2)
+    masks = masks.transpose((2, 0, 1)).astype(np.uint8)
+    bboxes = _coco_2_pascal(bboxes)
+
+    transformed = rand_crop(
+        image=img,
+        masks=list(masks),
+        bboxes=bboxes,
+        bbox_ids=np.arange(len(bboxes)),
+        labels=labels,
+    )
+
+    img = transformed["image"].astype(np.uint8)
+    shape = img.shape
+    labels = np.array(transformed["labels"], dtype=np.int64)
+    bboxes = np.array(transformed["bboxes"], dtype=np.float32).reshape(-1, 4)
+
+    bbox_ids = transformed["bbox_ids"]
+    masks = [transformed["masks"][i] for i in transformed["bbox_ids"]]
+    if masks:
+        masks = np.stack(masks, axis=0).astype(bool)
+    else:
+        masks = np.empty((0, shape[0], shape[1]), dtype=bool)
+
+    return pipeline(
+        {
+            "img": img,
+            "img_fields": ["img"],
+            "filename": None,
+            "ori_filename": None,
+            "img_shape": shape,
+            "ori_shape": shape,
+            "gt_masks": BitmapMasks(masks, *shape[:2]),
+            "gt_bboxes": bboxes,
+            "gt_labels": labels,
+            "bbox_fields": ["gt_bboxes"],
+        }
+    )
+
+
```

Comment:
This method is too long, we need to break down it, possibly create a wrapper class, and then use it to make this method smaller.  We need to increase code coherence here. Every function needs to do only 1 thing at most 2 or 3.

For instance like this:

```suggestion
class DataloaderBuilder:
    def __init__(
        self, 
        ds,
        cfg, 
        images_tensor, 
        masks_tensor, 
        boxes_tensor, 
        labels_tensor, 
        **train_dataloader_config
    ):
        self.pipeline = build_pipeline(cfg.train_pipeline)
        self.images_tensor = images_tensor or _find_tensor_with_htype(ds, "image")
        self.masks_tensor = masks_tensor or _find_tensor_with_htype(ds, "binary_mask")
        self.boxes_tensor = boxes_tensor or _find_tensor_with_htype(ds, "bbox")
        self.labels_tensor = labels_tensor or _find_tensor_with_htype(ds, "class_label")
        self.transform = self._transform
        self.num_workers, self.shuffle = self._parse_config()
    
    def _transform(self):
        return partial(
             self.transform,
             images_tensor=self.images_tensor,
             masks_tensor=self.masks_tensor,
             boxes_tensor=self.boxes_tensor,
             labels_tensor=self.labels_tensor,
             pipeline=self.pipeline,
         )
        
    def _parse_config(self):
        num_workers = self.train_loader_config["workers_per_gpu"]
        shuffle = self.train_loader_config.get("shuffle", True)
        return num_workers, shuffle
        
    def build_dataloader(self):
        dataloader = self.ds.pytorch(
             num_workers=self.num_workers,
             shuffle=self.shuffle,
             transform=self.transform_fn,
             tensors=[self.images_tensor, self.labels_tensor, self.boxes_tensor, self.masks_tensor],
             collate_fn=partial(
                 collate, samples_per_gpu=self.train_loader_config["samples_per_gpu"]
             ),
         )
         
         self._set_dataloader_classes(dataloader)
         return dataloader
         
     @staticmethod    
     def _set_dataloader_classes(dataloader):
         dataloader.dataset.CLASSES = [
             c["name"] for c in ds.categories.info["category_info"]
         ]
```